File: apps/web/src/lib/nutrition/meal_template.test.ts
import { test } from 'node:test';
import { strict as assert } from 'node:assert';
import {
	templateFromEntries,
	entriesFromTemplate,
	type TemplateSourceEntry,
	type PlannedTemplate,
	type PlannedTemplateItem,
} from './meal_template';

function entry(
	item_name: string,
	opts: Partial<TemplateSourceEntry> = {},
): TemplateSourceEntry {
	return { item_name, ...opts };
}

function pitem(
	position: number,
	itemName: string,
	opts: Partial<PlannedTemplateItem> = {},
): PlannedTemplateItem {
	return {
		position,
		itemName,
		mealSlot: null,
		calories: null,
		proteinG: null,
		carbsG: null,
		fatG: null,
		externalId: null,
		...opts,
	};
}

// ── templateFromEntries ─────────────────────────────────────────────────────

test('templateFromEntries copies name, item order, and macros', () => {
	const d = templateFromEntries('Pre-run breakfast', [
		entry('Oats', { calories: 300, protein_g: 10, carbs_g: 54, fat_g: 6 }),
		entry('Banana', { calories: 105, carbs_g: 27 }),
	]);
	assert.equal(d.name, 'Pre-run breakfast');
	assert.equal(d.itemCount, 2);
	assert.equal(d.items.length, 2);
	assert.deepEqual(
		d.items.map((i) => [i.position, i.itemName]),
		[[0, 'Oats'], [1, 'Banana']],
	);
	assert.equal(d.items[0].calories, 300);
	assert.equal(d.items[0].proteinG, 10);
	assert.equal(d.items[1].carbsG, 27);
	assert.equal(d.items[1].calories, 105);
	assert.equal(d.items[1].proteinG, null);
});

test('templateFromEntries falls back to default name when blank', () => {
	assert.equal(templateFromEntries('', [entry('Egg')]).name, 'Meal');
	assert.equal(templateFromEntries('   ', [entry('Egg')]).name, 'Meal');
	assert.equal(templateFromEntries(null, [entry('Egg')]).name, 'Meal');
	assert.equal(templateFromEntries(undefined, [entry('Egg')]).name, 'Meal');
});

test('templateFromEntries honours a custom fallback name', () => {
	assert.equal(templateFromEntries('', [entry('Egg')], 'Lunch box').name, 'Lunch box');
});

test('templateFromEntries drops blank-named entries and re-indexes positions', () => {
	const d = templateFromEntries('x', [
		entry('Oats'),
		entry('   '),
		entry(''),
		entry('Coffee'),
	]);
	assert.equal(d.itemCount, 2);
	assert.deepEqual(
		d.items.map((i) => [i.position, i.itemName]),
		[[0, 'Oats'], [1, 'Coffee']],
	);
});

test('templateFromEntries trims item names', () => {
	const d = templateFromEntries('x', [entry('  Greek yogurt  ')]);
	assert.equal(d.items[0].itemName, 'Greek yogurt');
});

test('templateFromEntries default slot = common slot when all agree', () => {
	const d = templateFromEntries('x', [
		entry('Oats', { meal_slot: 'breakfast' }),
		entry('Banana', { meal_slot: 'breakfast' }),
	]);
	assert.equal(d.mealSlot, 'breakfast');
});

test('templateFromEntries default slot = null when slots disagree', () => {
	const d = templateFromEntries('x', [
		entry('Oats', { meal_slot: 'breakfast' }),
		entry('Soup', { meal_slot: 'lunch' }),
	]);
	assert.equal(d.mealSlot, null);
});

test('templateFromEntries slotless entries do not veto a common slot', () => {
	const d = templateFromEntries('x', [
		entry('Oats', { meal_slot: 'dinner' }),
		entry('Water'),
		entry('Rice', { meal_slot: 'dinner' }),
	]);
	assert.equal(d.mealSlot, 'dinner');
});

test('templateFromEntries default slot = null when no entry has a slot', () => {
	const d = templateFromEntries('x', [entry('Oats'), entry('Banana')]);
	assert.equal(d.mealSlot, null);
});

test('templateFromEntries keeps per-item slot even when default is null', () => {
	const d = templateFromEntries('x', [
		entry('Oats', { meal_slot: 'breakfast' }),
		entry('Soup', { meal_slot: 'lunch' }),
	]);
	assert.equal(d.items[0].mealSlot, 'breakfast');
	assert.equal(d.items[1].mealSlot, 'lunch');
});

test('templateFromEntries ignores an invalid slot value', () => {
	const d = templateFromEntries('x', [
		entry('Oats', { meal_slot: 'brunch' as unknown as 'breakfast' }),
	]);
	assert.equal(d.items[0].mealSlot, null);
	assert.equal(d.mealSlot, null);
});

test('templateFromEntries coerces numeric strings and rejects NaN macros', () => {
	const d = templateFromEntries('x', [
		entry('Oats', {
			calories: '300' as unknown as number,
			protein_g: 'abc' as unknown as number,
		}),
	]);
	assert.equal(d.items[0].calories, 300);
	assert.equal(d.items[0].proteinG, null);
});

test('templateFromEntries preserves the Open Food Facts external_id', () => {
	const d = templateFromEntries('x', [entry('Oats', { external_id: 'off:123' })]);
	assert.equal(d.items[0].externalId, 'off:123');
});

test('templateFromEntries empty input yields an empty draft', () => {
	const d = templateFromEntries('Empty', []);
	assert.equal(d.itemCount, 0);
	assert.equal(d.items.length, 0);
	assert.equal(d.mealSlot, null);
});

// ── entriesFromTemplate ─────────────────────────────────────────────────────

function tpl(items: PlannedTemplateItem[], mealSlot: PlannedTemplate['mealSlot'] = null): PlannedTemplate {
	return { name: 'T', mealSlot, items };
}

test('entriesFromTemplate maps items to log inputs in position order', () => {
	const inputs = entriesFromTemplate(
		tpl([
			pitem(1, 'Banana', { calories: 105 }),
			pitem(0, 'Oats', { calories: 300, proteinG: 10 }),
		]),
	);
	assert.deepEqual(
		inputs.map((i) => i.itemName),
		['Oats', 'Banana'],
	);
	assert.equal(inputs[0].calories, 300);
	assert.equal(inputs[0].proteinG, 10);
	assert.equal(inputs[1].calories, 105);
});

test('entriesFromTemplate item slot wins over template + override', () => {
	const inputs = entriesFromTemplate(
		tpl([pitem(0, 'Oats', { mealSlot: 'breakfast' })], 'dinner'),
		'lunch',
	);
	assert.equal(inputs[0].mealSlot, 'breakfast');
});

test('entriesFromTemplate template default slot fills a slotless item', () => {
	const inputs = entriesFromTemplate(tpl([pitem(0, 'Oats')], 'dinner'), 'lunch');
	assert.equal(inputs[0].mealSlot, 'dinner');
});

test('entriesFromTemplate override fills when item + template slot are null', () => {
	const inputs = entriesFromTemplate(tpl([pitem(0, 'Oats')], null), 'snack');
	assert.equal(inputs[0].mealSlot, 'snack');
});

test('entriesFromTemplate slot is null when nothing supplies one', () => {
	const inputs = entriesFromTemplate(tpl([pitem(0, 'Oats')], null), null);
	assert.equal(inputs[0].mealSlot, null);
});

test('entriesFromTemplate carries macros + external_id through', () => {
	const inputs = entriesFromTemplate(
		tpl([
			pitem(0, 'Oats', {
				calories: 300,
				proteinG: 10,
				carbsG: 54,
				fatG: 6,
				externalId: 'off:42',
			}),
		]),
	);
	assert.deepEqual(inputs[0], {
		itemName: 'Oats',
		mealSlot: null,
		calories: 300,
		proteinG: 10,
		carbsG: 54,
		fatG: 6,
		externalId: 'off:42',
	});
});

test('entriesFromTemplate empty template yields no inputs', () => {
	assert.deepEqual(entriesFromTemplate(tpl([])), []);
});

test('entriesFromTemplate round-trips a saved-then-logged meal', () => {
	const draft = templateFromEntries('Lunch', [
		entry('Chicken', { meal_slot: 'lunch', calories: 250, protein_g: 40 }),
		entry('Rice', { meal_slot: 'lunch', calories: 200, carbs_g: 44 }),
	]);
	const planned: PlannedTemplate = {
		name: draft.name,
		mealSlot: draft.mealSlot,
		items: draft.items,
	};
	const inputs = entriesFromTemplate(planned);
	assert.equal(inputs.length, 2);
	assert.equal(inputs[0].itemName, 'Chicken');
	assert.equal(inputs[0].mealSlot, 'lunch');
	assert.equal(inputs[0].calories, 250);
	assert.equal(inputs[1].carbsG, 44);
});

File: apps/mobile_android/test/meal_template_test.dart
import 'package:flutter_test/flutter_test.dart';

import '../lib/meal_template.dart';

TemplateSourceEntry entry(
  String itemName, {
  String? mealSlot,
  double? calories,
  double? proteinG,
  double? carbsG,
  double? fatG,
  String? externalId,
}) =>
    TemplateSourceEntry(
      itemName: itemName,
      mealSlot: mealSlot,
      calories: calories,
      proteinG: proteinG,
      carbsG: carbsG,
      fatG: fatG,
      externalId: externalId,
    );

PlannedTemplateItem pitem(
  int position,
  String itemName, {
  String? mealSlot,
  double? calories,
  double? proteinG,
  double? carbsG,
  double? fatG,
  String? externalId,
}) =>
    PlannedTemplateItem(
      position: position,
      itemName: itemName,
      mealSlot: mealSlot,
      calories: calories,
      proteinG: proteinG,
      carbsG: carbsG,
      fatG: fatG,
      externalId: externalId,
    );

PlannedTemplate tpl(List<PlannedTemplateItem> items, {String? mealSlot}) =>
    PlannedTemplate(name: 'T', mealSlot: mealSlot, items: items);

void main() {
  // ── templateFromEntries ───────────────────────────────────────────────────

  test('templateFromEntries copies name, item order, and macros', () {
    final d = templateFromEntries('Pre-run breakfast', [
      entry('Oats', calories: 300, proteinG: 10, carbsG: 54, fatG: 6),
      entry('Banana', calories: 105, carbsG: 27),
    ]);
    expect(d.name, 'Pre-run breakfast');
    expect(d.itemCount, 2);
    expect(d.items.length, 2);
    expect(d.items.map((i) => [i.position, i.itemName]).toList(),
        [[0, 'Oats'], [1, 'Banana']]);
    expect(d.items[0].calories, 300);
    expect(d.items[0].proteinG, 10);
    expect(d.items[1].carbsG, 27);
    expect(d.items[1].calories, 105);
    expect(d.items[1].proteinG, null);
  });

  test('templateFromEntries falls back to default name when blank', () {
    expect(templateFromEntries('', [entry('Egg')]).name, 'Meal');
    expect(templateFromEntries('   ', [entry('Egg')]).name, 'Meal');
    expect(templateFromEntries(null, [entry('Egg')]).name, 'Meal');
  });

  test('templateFromEntries honours a custom fallback name', () {
    expect(
        templateFromEntries('', [entry('Egg')], fallbackName: 'Lunch box').name,
        'Lunch box');
  });

  test('templateFromEntries drops blank-named entries and re-indexes positions',
      () {
    final d = templateFromEntries('x', [
      entry('Oats'),
      entry('   '),
      entry(''),
      entry('Coffee'),
    ]);
    expect(d.itemCount, 2);
    expect(d.items.map((i) => [i.position, i.itemName]).toList(),
        [[0, 'Oats'], [1, 'Coffee']]);
  });

  test('templateFromEntries trims item names', () {
    final d = templateFromEntries('x', [entry('  Greek yogurt  ')]);
    expect(d.items[0].itemName, 'Greek yogurt');
  });

  test('templateFromEntries default slot = common slot when all agree', () {
    final d = templateFromEntries('x', [
      entry('Oats', mealSlot: 'breakfast'),
      entry('Banana', mealSlot: 'breakfast'),
    ]);
    expect(d.mealSlot, 'breakfast');
  });

  test('templateFromEntries default slot = null when slots disagree', () {
    final d = templateFromEntries('x', [
      entry('Oats', mealSlot: 'breakfast'),
      entry('Soup', mealSlot: 'lunch'),
    ]);
    expect(d.mealSlot, null);
  });

  test('templateFromEntries slotless entries do not veto a common slot', () {
    final d = templateFromEntries('x', [
      entry('Oats', mealSlot: 'dinner'),
      entry('Water'),
      entry('Rice', mealSlot: 'dinner'),
    ]);
    expect(d.mealSlot, 'dinner');
  });

  test('templateFromEntries default slot = null when no entry has a slot', () {
    final d = templateFromEntries('x', [entry('Oats'), entry('Banana')]);
    expect(d.mealSlot, null);
  });

  test('templateFromEntries keeps per-item slot even when default is null', () {
    final d = templateFromEntries('x', [
      entry('Oats', mealSlot: 'breakfast'),
      entry('Soup', mealSlot: 'lunch'),
    ]);
    expect(d.items[0].mealSlot, 'breakfast');
    expect(d.items[1].mealSlot, 'lunch');
  });

  test('templateFromEntries ignores an invalid slot value', () {
    final d = templateFromEntries('x', [entry('Oats', mealSlot: 'brunch')]);
    expect(d.items[0].mealSlot, null);
    expect(d.mealSlot, null);
  });

  test('templateFromEntries coerces numeric strings and rejects NaN macros',
      () {
    final d = templateFromEntries('x', [
      TemplateSourceEntry(
        itemName: 'Oats',
        calories: null,
        proteinG: double.nan,
      ),
    ]);
    expect(d.items[0].proteinG, null);
  });

  test('templateFromEntries preserves the Open Food Facts external_id', () {
    final d = templateFromEntries('x', [entry('Oats', externalId: 'off:123')]);
    expect(d.items[0].externalId, 'off:123');
  });

  test('templateFromEntries empty input yields an empty draft', () {
    final d = templateFromEntries('Empty', []);
    expect(d.itemCount, 0);
    expect(d.items.length, 0);
    expect(d.mealSlot, null);
  });

  // ── entriesFromTemplate ───────────────────────────────────────────────────

  test('entriesFromTemplate maps items to log inputs in position order', () {
    final inputs = entriesFromTemplate(tpl([
      pitem(1, 'Banana', calories: 105),
      pitem(0, 'Oats', calories: 300, proteinG: 10),
    ]));
    expect(inputs.map((i) => i.itemName).toList(), ['Oats', 'Banana']);
    expect(inputs[0].calories, 300);
    expect(inputs[0].proteinG, 10);
    expect(inputs[1].calories, 105);
  });

  test('entriesFromTemplate item slot wins over template + override', () {
    final inputs = entriesFromTemplate(
      tpl([pitem(0, 'Oats', mealSlot: 'breakfast')], mealSlot: 'dinner'),
      slotOverride: 'lunch',
    );
    expect(inputs[0].mealSlot, 'breakfast');
  });

  test('entriesFromTemplate template default slot fills a slotless item', () {
    final inputs = entriesFromTemplate(
      tpl([pitem(0, 'Oats')], mealSlot: 'dinner'),
      slotOverride: 'lunch',
    );
    expect(inputs[0].mealSlot, 'dinner');
  });

  test('entriesFromTemplate override fills when item + template slot are null',
      () {
    final inputs = entriesFromTemplate(
      tpl([pitem(0, 'Oats')]),
      slotOverride: 'snack',
    );
    expect(inputs[0].mealSlot, 'snack');
  });

  test('entriesFromTemplate slot is null when nothing supplies one', () {
    final inputs = entriesFromTemplate(tpl([pitem(0, 'Oats')]));
    expect(inputs[0].mealSlot, null);
  });

  test('entriesFromTemplate carries macros + external_id through', () {
    final inputs = entriesFromTemplate(tpl([
      pitem(0, 'Oats',
          calories: 300,
          proteinG: 10,
          carbsG: 54,
          fatG: 6,
          externalId: 'off:42'),
    ]));
    expect(inputs[0].itemName, 'Oats');
    expect(inputs[0].mealSlot, null);
    expect(inputs[0].calories, 300);
    expect(inputs[0].proteinG, 10);
    expect(inputs[0].carbsG, 54);
    expect(inputs[0].fatG, 6);
    expect(inputs[0].externalId, 'off:42');
  });

  test('entriesFromTemplate empty template yields no inputs', () {
    expect(entriesFromTemplate(tpl([])), isEmpty);
  });

  test('entriesFromTemplate round-trips a saved-then-logged meal', () {
    final draft = templateFromEntries('Lunch', [
      entry('Chicken', mealSlot: 'lunch', calories: 250, proteinG: 40),
      entry('Rice', mealSlot: 'lunch', calories: 200, carbsG: 44),
    ]);
    final planned = PlannedTemplate(
      name: draft.name,
      mealSlot: draft.mealSlot,
      items: [
        for (final it in draft.items)
          pitem(it.position, it.itemName,
              mealSlot: it.mealSlot,
              calories: it.calories,
              proteinG: it.proteinG,
              carbsG: it.carbsG,
              fatG: it.fatG,
              externalId: it.externalId),
      ],
    );
    final inputs = entriesFromTemplate(planned);
    expect(inputs.length, 2);
    expect(inputs[0].itemName, 'Chicken');
    expect(inputs[0].mealSlot, 'lunch');
    expect(inputs[0].calories, 250);
    expect(inputs[1].carbsG, 44);
  });
}
